src/shared/frame-rpc.js

```diff
@@ -53,46 +60,67 @@ const VERSION = '1.0.0';
 export class RPC {
   /**
    * Create an RPC client for sending RPC requests from `sourceFrame` to
-   * `destFrame`.
+   * `destFrame`, and receiving RPC responses from `destFrame` to `sourceFrame`.
    *
-   * @param {Window} sourceFrame
-   * @param {Window} destFrame
-   * @param {string} origin - Origin of destination frame
-   * @param {Record<string, (...args: any[]) => any>} methods - Map of method
+   * This class has been adapted to work with `MessageChannel`. Because
+   * `MessageChannel` are unidirectional we ignore the `sourceFrame` and `origin`
+   * properties, in this case.
+   *
+   * TODO: July 2021, currently this class is a bit of a _frankenstein_ because
+   * we support both, `Window.postMessage` and `MessagePort.postMessage`.
+   * Once we move all the inter-frame communication to `MessageChannel` we will
+   * be able to cleanup this class. We have added `@deprecated` statements in
+   * the pieces of code that need to be removed.
+   *
+   * @param {Window} sourceFrame -- @deprecated
+   * @param {Window|MessagePort} destFrameOrPort
+   * @param {string} origin - Origin of destination frame @deprecated
+   * @param {Record<string, (...args: any[]) => void>} methods - Map of method
    *   name to method handler
    */
-  constructor(sourceFrame, destFrame, origin, methods) {
-    this.sourceFrame = sourceFrame;
-    this.destFrame = destFrame;
+  constructor(sourceFrame, destFrameOrPort, origin, methods) {
```

Comment:
Since we're only switching from one implementation to another, if/else statements are a reasonable way to go. An alternative migration strategy might have been to wrap the existing `window.postMessage` logic up in a `MessagePort`-like interface adapter. This would isolate all the legacy `window.postMessage` logic in one place.

An example:

<details>

```diff
diff --git a/src/shared/bridge.js b/src/shared/bridge.js
index 195ff80e5..a60a49715 100644
--- a/src/shared/bridge.js
+++ b/src/shared/bridge.js
@@ -1,4 +1,4 @@
-import { RPC } from './frame-rpc';
+import { RPC, WindowPort } from './frame-rpc';
 
 /** @typedef {import('../types/annotator').Destroyable} Destroyable */
 
@@ -59,7 +59,8 @@ export default class Bridge {
     const listeners = { connect, ...this.channelListeners };
 
     // Set up a channel
-    channel = new RPC(window, source, origin, listeners);
+    const port = new WindowPort(window, source, origin);
+    channel = new RPC(port, listeners);
 
     // Fire off a connection attempt
     channel.call('connect', token, ready);
diff --git a/src/shared/frame-rpc.js b/src/shared/frame-rpc.js
index 72780ab2a..1314215fb 100644
--- a/src/shared/frame-rpc.js
+++ b/src/shared/frame-rpc.js
@@ -1,5 +1,3 @@
-import { ListenerCollection } from '../annotator/util/listener-collection';
-
 /*
   This module was adapted from `index.js` in https://github.com/substack/frame-rpc.
 
@@ -50,6 +48,53 @@ const PROTOCOL = 'frame-rpc';
  * @typedef {import('../types/annotator').Destroyable} Destroyable
  */
 
+/**
+ * Implementation of the `MessagePort` interface that uses `window.postMessage`.
+ *
+ * This exists to allow the `RPC` class to be migrated from working with
+ * `window.postMessage` to `MessagePort`.
+ */
+export class WindowPort {
+  /**
+   * @param {Window} sourceWindow
+   * @param {Window} destWindow
+   * @param {string} origin
+   */
+  constructor(sourceWindow, destWindow, origin) {
+    this.sourceWindow = sourceWindow;
+    this.destWindow = destWindow;
+    this.origin = new URL(origin).origin;
+
+    this._listener = null;
+    this._handleMessage = event => {
+      if (
+        (origin !== '*' && event.origin !== origin) ||
+        event.source !== destWindow
+      ) {
+        return;
+      }
+      this._listener?.(event);
+    };
+    this.sourceWindow.addEventListener('message', this._handleMessage);
+  }
+
+  /** @param {(e: MessageEvent) => void} handler */
+  set onmessage(handler) {
+    this._listener = handler;
+  }
+
+  close() {
+    this.sourceWindow.removeEventListener('message', this._handleMessage);
+  }
+
+  /**
+   * @param {any} message
+   */
+  postMessage(message) {
+    this.destWindow.postMessage(message, this.origin);
+  }
+}
+
 /**
  * Class for making RPC requests between two frames.
  *
@@ -72,28 +117,12 @@ export class RPC {
    * be able to cleanup this class. We have added `@deprecated` statements in
    * the pieces of code that need to be removed.
    *
-   * @param {Window} sourceFrame -- @deprecated
-   * @param {Window|MessagePort} destFrameOrPort
-   * @param {string} origin - Origin of destination frame @deprecated
+   * @param {WindowPort|MessagePort} port
    * @param {Record<string, (...args: any[]) => void>} methods - Map of method
    *   name to method handler
    */
-  constructor(sourceFrame, destFrameOrPort, origin, methods) {
-    this.sourceFrame = sourceFrame; // sourceFrame is ignored if using MessagePort
-
-    if (destFrameOrPort instanceof MessagePort) {
-      this._port = destFrameOrPort;
-    } else {
-      /** @deprecated */
-      this.destFrame = destFrameOrPort;
-    }
-
-    /** @deprecated */
-    if (origin === '*') {
-      this.origin = '*';
-    } else {
-      this.origin = new URL(origin).origin;
-    }
+  constructor(port, methods) {
+    this._port = port;
 
     this._methods = methods;
 
@@ -101,26 +130,8 @@ export class RPC {
     this._callbacks = {};
     this._destroyed = false;
 
-    this._listeners = new ListenerCollection();
-
-    if (this._port) {
-      this._listeners.add(this._port, 'message', event =>
-        this._handle(/** @type {MessageEvent} */ (event))
-      );
-      this._port.start();
-    } else {
-      /** @deprecated */
-      /** @param {MessageEvent} event */
-      const onmessage = event => {
-        if (!this._isValidSender) {
-          return;
-        }
-        this._handle(event);
-      };
-      this._listeners.add(this.sourceFrame, 'message', event =>
-        onmessage(/** @type {MessageEvent} */ (event))
-      );
-    }
+    this._handle = this._handle.bind(this);
+    this._port.onmessage = this._handle;
   }
 
   /**
@@ -129,10 +140,7 @@ export class RPC {
    */
   destroy() {
     this._destroyed = true;
-    this._listeners.removeAll();
-    if (this._port) {
-      this._port.close();
-    }
+    this._port.close();
   }
 
   /**
@@ -165,31 +173,7 @@ export class RPC {
       version: VERSION,
     };
 
-    if (this._port) {
-      this._port.postMessage(message);
-    }
-
-    /** @deprecated */
-    if (this.destFrame) {
-      this.destFrame.postMessage(message, this.origin);
-    }
-  }
-
-  /**
-   * Validate sender
-   *
-   * @param {MessageEvent} event
-   * @deprecated
-   */
-  _isValidSender(event) {
-    if (this.destFrame !== event.source) {
-      return false;
-    }
-    if (this.origin !== '*' && event.origin !== this.origin) {
-      return false;
-    }
-
-    return true;
+    this._port.postMessage(message);
   }
 
   /**
@@ -239,14 +223,7 @@ export class RPC {
           version: VERSION,
         };
 
-        if (this._port) {
-          this._port.postMessage(message);
-        }
-
-        /** @deprecated */
-        if (this.destFrame) {
-          this.destFrame.postMessage(message, this.origin);
-        }
+        this._port.postMessage(message);
       };
       this._methods[msg.method].call(this._methods, ...msg.arguments, callback);
     } else if ('response' in msg) {
```

</details>

A more lightweight change compared to this PR might have been to bundle up the properties that related to the old communication method in one object, such that the `_port` field refers either to the `MessagePort` or the object containing the source/dest window and the origin. For example:

<details>

```diff
diff --git a/src/shared/bridge.js b/src/shared/bridge.js
index 195ff80e5..78e6ba679 100644
--- a/src/shared/bridge.js
+++ b/src/shared/bridge.js
@@ -59,7 +59,7 @@ export default class Bridge {
     const listeners = { connect, ...this.channelListeners };
 
     // Set up a channel
-    channel = new RPC(window, source, origin, listeners);
+    channel = new RPC({ source: window, dest: source, origin }, listeners);
 
     // Fire off a connection attempt
     channel.call('connect', token, ready);
diff --git a/src/shared/frame-rpc.js b/src/shared/frame-rpc.js
index 72780ab2a..469c83bff 100644
--- a/src/shared/frame-rpc.js
+++ b/src/shared/frame-rpc.js
@@ -50,6 +50,13 @@ const PROTOCOL = 'frame-rpc';
  * @typedef {import('../types/annotator').Destroyable} Destroyable
  */
 
+/**
+ * @typedef MessageSource
+ * @prop {Window} dest
+ * @prop {Window} source
+ * @prop {string} origin
+ */
+
 /**
  * Class for making RPC requests between two frames.
  *
@@ -72,27 +79,15 @@ export class RPC {
    * be able to cleanup this class. We have added `@deprecated` statements in
    * the pieces of code that need to be removed.
    *
-   * @param {Window} sourceFrame -- @deprecated
-   * @param {Window|MessagePort} destFrameOrPort
-   * @param {string} origin - Origin of destination frame @deprecated
+   * @param {MessageSource|MessagePort} port
    * @param {Record<string, (...args: any[]) => void>} methods - Map of method
    *   name to method handler
    */
-  constructor(sourceFrame, destFrameOrPort, origin, methods) {
-    this.sourceFrame = sourceFrame; // sourceFrame is ignored if using MessagePort
+  constructor(port, methods) {
+    this._port = port;
 
-    if (destFrameOrPort instanceof MessagePort) {
-      this._port = destFrameOrPort;
-    } else {
-      /** @deprecated */
-      this.destFrame = destFrameOrPort;
-    }
-
-    /** @deprecated */
-    if (origin === '*') {
-      this.origin = '*';
-    } else {
-      this.origin = new URL(origin).origin;
+    if (!(port instanceof MessagePort)) {
+      port.origin = new URL(port.origin).origin;
     }
 
     this._methods = methods;
@@ -103,21 +98,24 @@ export class RPC {
 
     this._listeners = new ListenerCollection();
 
-    if (this._port) {
+    if (this._port instanceof MessagePort) {
       this._listeners.add(this._port, 'message', event =>
         this._handle(/** @type {MessageEvent} */ (event))
       );
       this._port.start();
     } else {
-      /** @deprecated */
+      const port = this._port;
       /** @param {MessageEvent} event */
       const onmessage = event => {
-        if (!this._isValidSender) {
+        if (port.dest !== event.source) {
+          return;
+        }
+        if (port.origin !== '*' && event.origin !== port.origin) {
           return;
         }
         this._handle(event);
       };
-      this._listeners.add(this.sourceFrame, 'message', event =>
+      this._listeners.add(this._port.dest, 'message', event =>
         onmessage(/** @type {MessageEvent} */ (event))
       );
     }
@@ -130,7 +128,7 @@ export class RPC {
   destroy() {
     this._destroyed = true;
     this._listeners.removeAll();
-    if (this._port) {
+    if (this._port instanceof MessagePort) {
       this._port.close();
     }
   }
@@ -165,31 +163,11 @@ export class RPC {
       version: VERSION,
     };
 
-    if (this._port) {
+    if (this._port instanceof MessagePort) {
       this._port.postMessage(message);
+    } else {
+      this._port.dest.postMessage(message, this._port.origin);
     }
-
-    /** @deprecated */
-    if (this.destFrame) {
-      this.destFrame.postMessage(message, this.origin);
-    }
-  }
-
-  /**
-   * Validate sender
-   *
-   * @param {MessageEvent} event
-   * @deprecated
-   */
-  _isValidSender(event) {
-    if (this.destFrame !== event.source) {
-      return false;
-    }
-    if (this.origin !== '*' && event.origin !== this.origin) {
-      return false;
-    }
-
-    return true;
   }
 
   /**
@@ -239,13 +217,10 @@ export class RPC {
           version: VERSION,
         };
 
-        if (this._port) {
+        if (this._port instanceof MessagePort) {
           this._port.postMessage(message);
-        }
-
-        /** @deprecated */
-        if (this.destFrame) {
-          this.destFrame.postMessage(message, this.origin);
+        } else {
+          this._port.dest.postMessage(message, this._port.origin);
         }
       };
       this._methods[msg.method].call(this._methods, ...msg.arguments, callback);
diff --git a/src/shared/test/frame-rpc-test.js b/src/shared/test/frame-rpc-test.js
index 2d24ac10f..6765baafa 100644
--- a/src/shared/test/frame-rpc-test.js
+++ b/src/shared/test/frame-rpc-test.js
@@ -19,14 +19,9 @@ describe('shared/bridge', () => {
       callback(result);
     };
 
-    rpc1 = new RPC(
-      /** dummy when using ports */ window,
-      port1,
-      /** dummy when using ports */ '*',
-      {
-        concat,
-      }
-    );
+    rpc1 = new RPC(port1, {
+      concat,
+    });
 
     // `plusOne` method for rpc2
     plusOne = sinon.stub().callsFake((...numbers) => {
@@ -35,14 +30,9 @@ describe('shared/bridge', () => {
       callback(result);
     });
 
-    rpc2 = new RPC(
-      /** dummy when using ports */ window,
-      port2,
-      /** dummy when using ports */ '*',
-      {
-        plusOne,
-      }
-    );
+    rpc2 = new RPC(port2, {
+      plusOne,
+    });
   });
 
   afterEach(() => {
```

</details>

Reply:
Thank you very much for the recommendation. I did think about doing some similar to what you suggested.

I thought that removing the few branches that are marked as deprecated would have been easier.

An important  aspect is that I want to support `MessagePort` in `RPC` and `Bridge` maintaining the support for `Window`. In case of unexpected issues I want to be able to rollback easily.